proposals/nnnn-safe-shared-contiguous-storage.md

```diff
@@ -0,0 +1,1232 @@
+3. It might escape the duration of the closure
+
+Even if the body of the `withUnsafeXXX` call does not escape the pointer, other functions called inside the closure have to be written in terms of unsafe pointers. This requires programmer vigilance across a project and potentially spreads the use of unsafe types, even when it could have been written in terms of safe constructs.
+
+We want to take advantage of the features of non-escapable types to replace some closure-taking API with simple properties, resulting in more composable code:
+
+```swift
+let array = Array("Hello\0".utf8)
+array.withUnsafeBufferPointer {
+  // use `$0` here for direct memory access
+}
+let span: Span<UInt8> = array.storage
+// use `span` in the same scope as `array` for direct memory access
+```
+
+## Proposed solution
+
+`Span` will allow sharing the contiguous internal representation of a type, by providing access to a borrowed view of an interval of contiguous memory. A view does not copy the underlying data: it instead relies on a guarantee that the original container cannot be modified or destroyed during the lifetime of the view. `Span`'s lifetime is statically enforced as a lifetime dependency to a binding of the type vending it, preventing its escape from the scope where it is valid for use. This guarantee preserves temporal safety. `Span` also performs bounds-checking on every access to preserve spatial safety. Additionally `Span` always represents initialized memory, preserving the definite initialization guarantee.
+
+By relying on borrowing, `Span` can provide simultaneous access to a non-copyable container, and can help avoid unwanted copies of copyable containers. Note that `Span` is not a replacement for a copyable container with owned storage; see the future directions for more details ([Resizable, contiguously-stored, untyped collection in the standard library](#Bytes))
+
+`Span` is the currency type for local processing over values in contiguous memory. It is the replacement for any API currently using `Array`, `UnsafeBufferPointer`, `Foundation.Data`, etc., that does not need to escape the value.
+
+### `ContiguousStorage`
+
+A type can indicate that it can provide a `Span` by conforming to the `ContiguousStorage` protocol. `ContiguousStorage` forms a bridge between multi-type or generically-typed interfaces and a performant concrete implementation.
+
+For example, for the hypothetical base64 decoding library mentioned above, a possible API could be:
+
+```swift
+extension HypotheticalBase64Decoder {
+  public func decode(bytes: some ContiguousStorage<UInt8>) -> [UInt8]
+}
+```
+
+Even better, an interface can be defined in terms of the concrete type `Span<UInt8>`:
+
+```swift
+extension Hypothetical Base64Decoder {
+  public func decode(bytes: Span<UInt8>) -> [UInt8]
+}
+```
+
+Advanced libraries might add use an inlinable generic-dispatch interface in addition to a concrete interface defined in terms of `Span`
```

Comment:
We say "even better" for `Span` vs `ContiguouslyStored`, but don't directly prescribe this. We can insert some of the prose that recommends using `Span` in most places and leaving `ContiguouslyStored` as adapter code or in select top-level API. Since `ContiguouslyStored` exposes a `.storage` property, it's not a huge obstacle.